src/giftCards/GiftCardUpdatePage/providers/GiftCardUpdateFormProvider.tsx

```diff
@@ -0,0 +1,123 @@
+import { MutationResultWithOpts } from "@saleor/hooks/makeMutation";
+import { UseFormResult } from "@saleor/hooks/useForm";
+import useNotifier from "@saleor/hooks/useNotifier";
+import { getDefaultNotifierSuccessErrorData } from "@saleor/hooks/useNotifier/utils";
+import { getFormErrors } from "@saleor/utils/errors";
+import React, { createContext, useContext } from "react";
+import { useIntl } from "react-intl";
+
+import { initialData as emptyFormData } from "../../GiftCardCreateDialog/GiftCardCreateDialogForm";
+import { useGiftCardUpdateMutation } from "../mutations";
+import { GiftCardUpdate } from "../types/GiftCardUpdate";
+import { GiftCardDetailsContext } from "./GiftCardDetailsProvider";
+
+interface GiftCardUpdateFormProviderProps {
+  children: React.ReactNode;
+  onBalanceUpdateSuccess: () => void;
+}
+
+export type GiftCardUpdateFormData = Omit<
+  GiftCardCommonFormData,
+  "balanceAmount" | "balanceCurrency"
+>;
+
+export interface GiftCardUpdateFormConsumerData
+  extends GiftCardUpdateFormErrors {
+  opts: MutationResultWithOpts<GiftCardUpdate>;
+}
+
+export interface GiftCardUpdateFormErrors {
+  formErrors: Record<"tag" | "expiryDate" | "expiryPeriod", GiftCardError>;
+}
+
+export type GiftCardUpdateFormConsumerProps = UseFormResult<
+  GiftCardUpdateFormData
+> &
+  GiftCardUpdateFormConsumerData;
+
+export const GiftCardUpdateFormContext = createContext<
+  GiftCardUpdateFormConsumerProps
+>(null);
+
+const GiftCardUpdateFormProvider: React.FC<GiftCardUpdateFormProviderProps> = ({
+  children,
+  onBalanceUpdateSuccess
+}) => {
+  const notify = useNotifier();
+  const intl = useIntl();
+
+  const { loading: loadingGiftCard, giftCard } = useContext(
+    GiftCardDetailsContext
+  );
+
+  const getInitialData = (): GiftCardUpdateFormData => {
+    if (loadingGiftCard || !giftCard) {
+      return emptyFormData;
+    }
+
+    const { tag, expiryDate, expiryType, expiryPeriod } = giftCard;
+
+    return {
+      tag,
+      expiryDate,
+      expiryType,
+      expiryPeriodType: expiryPeriod?.type,
+      expiryPeriodAmount: expiryPeriod?.amount
+    };
+  };
+
+  const onSubmit = (data: GiftCardUpdate) => {
+    const errors = data.giftCardUpdate.errors;
+
+    notify(getDefaultNotifierSuccessErrorData(errors, intl));
+
+    if (!errors.length) {
+      onBalanceUpdateSuccess();
+    }
+  };
+
+  const [updateGiftCard, updateGiftCardOpts] = useGiftCardUpdateMutation({
+    onCompleted: onSubmit
+  });
+
+  const handleSubmit = (formData: GiftCardUpdateFormData) => {
+    updateGiftCard({
+      variables: {
+        id: giftCard?.id,
+        input: {
+          tag: formData.tag,
+          expirySettings: getGiftCardExpirySettingsInputData(formData)
+        }
+      }
+    });
+  };
+
+  const formErrors = getFormErrors(
+    ["tag", "expiryDate", "expiryPeriod"],
+    updateGiftCardOpts?.data?.giftCardUpdate?.errors
+  );
+
+  return (
+    <Form initial={getInitialData()} onSubmit={handleSubmit}>
```

Comment:
This component will be rerendered quite often, see if there are any noticeable performance issues. In general all providers that do not immediately return `children` suffer from a number of rerenders. A quick fix is to `memo` the `Form` component and memoize the props to keep static references. `getInitialData` could be a useMemo and `handleSubmit` could use useCallback. See how often `Form` rerenders here by placing a `console.log` and count how many times it's called.

Reply:
So far nothing seems to be out of order, going to keep an eye on this form in the incoming pr's.